src/PropertySuggester/Suggesters/SimpleSuggester.php

```diff
@@ -68,40 +102,96 @@ protected function getSuggestions( array $propertyIds, array $idTuples, $limit,
 		if ( !is_float( $minProbability ) ) {
 			throw new InvalidArgumentException( '$minProbability must be float!' );
 		}
-		if ( !$propertyIds ) {
-			return array();
-		}
-
-		$excludedIds = array_merge( $propertyIds, $this->deprecatedPropertyIds );
-		$count = count( $propertyIds );
 
 		$dbr = $this->lb->getConnection( DB_SLAVE );
-		if ( empty( $idTuples ) ){
-			$condition = 'pid1 IN (' . $dbr->makeList( $propertyIds ) . ')';
-		}
-		else{
-			$condition = $dbr->makeList( $idTuples, LIST_OR );
-		}
+		$count = $this->getPropertyCount( $dbr, $propertyIds );
+
 		$res = $dbr->select(
 			'wbs_propertypairs',
 			array( 'pid' => 'pid2', 'prob' => "sum(probability)/$count" ),
-			array( $condition,
-				   'pid2 NOT IN (' . $dbr->makeList( $excludedIds ) . ')',
-				   'context' => $context ),
+			$this->makeGetSuggestionsConditions( $dbr, $propertyIds, $idTuples, $context ),
 			__METHOD__,
 			array(
 				'GROUP BY' => 'pid2',
 				'ORDER BY' => 'prob DESC',
-				'LIMIT'    => $limit,
+				'LIMIT'	=> $limit,
 				'HAVING'   => 'prob > ' . floatval( $minProbability )
-				)
-			);
+			)
+		);
+
 		$this->lb->reuseConnection( $dbr );
 
 		return $this->buildResult( $res );
 	}
 
 	/**
+	 * @param DatabaseBase $dbr
+	 * @param int[] $propertyIds
+	 *
+	 * @throws RuntimeException
+	 * @return int
+	 */
+	private function getPropertyCount( DatabaseBase $dbr, array $propertyIds ) {
+		$count = count( $propertyIds );
+
+		if ( $count === 0 ) {
+			$cacheKey = $this->cacheKeyPrefix . ':' . 'Suggester:PropertyCount';
+			$count = $this->cache->get( $cacheKey );
+
+			if ( $count === false ) {
+				$res = $dbr->selectRow(
+					'wbs_propertypairs',
+					'count(distinct(pid1)) as count',
+					array(),
+					__METHOD__
+				);
+
+				if ( !$res ) {
+					throw new RuntimeException( 'Failed to obtain property count.' );
+				}
+
+				$count = (int)$res->count;
+				$this->cache->set( $cacheKey, $count, $this->cacheDuration );
+			}
+		}
+
+		return $count;
+	}
+
+	/**
+	 * @param DatabaseBase $dbr
+	 * @param int[] $propertyIds
+	 * @param string[] $idTuples
+	 * @param string $context
+	 *
+	 * @return string[]
+	 */
+	private function makeGetSuggestionsConditions( DatabaseBase $dbr, array $propertyIds,
+		array $idTuples, $context
+	) {
+		$conditions = array(
+			'context' => $context
+		);
+
+		$excludedIds = array_merge( $propertyIds, $this->deprecatedPropertyIds );
```

Comment:
This doesn't make much sense (or I'm misreading the code): The same `$propertyIds` are excluded via `NOT IN` and included via `IN` just a few lines below, both in the same `$conditions`? Could you try to add comments or use more descriptive variable names?